### DISCUSSION TITLE
Preserve url params in paginate links

### DISCUSSION THREAD

**Discussion by bit-willi** [laravel/framework#49866] [Ideas] [answered]
When using Model::paginate() Laravel includes 'links' and 'meta' in respose. 
Example:  

GET - /users?name=John
```json
{
  "users": [
    {
      "id": 1,
      "name": "John Doe",
      "email": "john_doe@example.com",
      "password": "***",
      "created_at": "2021-01-01 00:00:00",
      "updated_at": "2021-01-01 00:00:00",
    }
  ],
  "links": {
    "first": "http://localhost:8000/api/users?page=1",
    "last": "http://localhost:8000/api/users?page=10",
    "prev": null,
    "next": "http://localhost:8000/api/users?page=2"
  },
  "meta": {
    "current_page": 1,
    "from": 1,
    "last_page": 1,
    "links": [
      {
        "url": "http://localhost:8000/api/users?page=1",
        "label": "1",
        "active": true
      }
    ],
    "path": "http://localhost:8000/api/users",
    "per_page": 15,
    "to": 1,
    "total": 10
  }
}
```

While building a REST API, we usually use query params to filter data in index(), the problem is that we lose these parameters when  use the `links.next`. I my opinion, query params should be preserved, if not automatically, at least we need a way to 'inject' the $request->validated() to preserve the filters in requests.

Example:
GET - /users?name=John
links.next should be 'https://localhost:8000/api/users?page=1&name=John'

**Comment by macropay-solutions** [comment]
Good point. We had to unset these links and handle the pagination in FE by changing the page there to avoid issues.
But bear in mind that long urls require increasing the request buffer size.

**Comment by hulkur** [comment, marked as answer]
withQueryString

`{{ $paginator->withQueryString()->links() }}`

**Reply by ghost** [reply]
$builder->paginate(); // LenghtAwarePaginator dows not have that function

**Reply by hulkur** [reply]
`class LengthAwarePaginator extends AbstractPaginator` and AbstractPaginator has `public function withQueryString()`

**Reply by ghost** [reply]
My IDE is stupid... You are right.

**Reply by hulkur** [reply]
police to lost driver: gps stupid, you driver! :)

**Reply by ghost** [reply]
If I call it like 
    
     paginate(...)->withQueryString() //is does not see it.

 If I put 

     /** @var LengthAwarePaginator $lengthAwarePaginator */

then it sees it even if the paginate has a dockblock of @return LengthAwarePaginator

**Comment by macropay-solutions** [comment]
@hulkur I wonder why this is not the default behavior...
It is worth mentioning that the query gathering can be customized before calling `withQueryString` with:

        Paginator::queryStringResolver(function () use ($app) {
            return $app['request']->query(); // change this
        });
